src/efftester.ml

```diff
@@ -297,7 +369,8 @@ let rec types_compat t t' =
 
 let imm_eff t =
   match t with
-  | Lit _ | Variable (_, _) | Lambda (_, _, _, _) -> no_eff
+  | Lit _ | Variable (_, _) | Constructor (_, _, _) | Lambda (_, _, _, _) -> no_eff
```

Comment:
Wait, shouldn't `Constructor` carry the effect of its arguments? Consider `Some (print_int 1)` for example. To me this suggests adding a last `eff` parameter to the `Constructor` declaration, as you did for `PatternMatch`, to be computed as the least upper bound of the effect of each of its arguments.